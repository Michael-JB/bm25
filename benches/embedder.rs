use bm25::{EmbedderBuilder, Language, LanguageMode, TokenEmbedder};
use divan::Bencher;
use std::hash::Hash;

use divan::AllocProfiler;

#[global_allocator]
static ALLOC: AllocProfiler = AllocProfiler::system();

fn main() {
    divan::main();
}

const ENGLISH_TEXT: &str =
    "The dominant sequence transduction models are based on complex recurrent or
convolutional neural networks in an encoder-decoder configuration. The best performing
models also connect the encoder and decoder through an attention mechanism. We
propose a new simple network architecture, the Transformer, based solely on attention
mechanisms, dispensing with recurrence and convolutions entirely. Experiments on two
machine translation tasks show these models to be superior in quality while being more
parallelizable and requiring significantly less time to train.";

#[divan::bench(args = [
    #[cfg(feature = "language_detection")]
    LanguageMode::Detect,
    LanguageMode::Fixed(Language::English),
], sample_count = 10000)]
fn language_mode(bencher: Bencher, language_mode: &LanguageMode) {
    let embedder =
        EmbedderBuilder::<u32>::with_fit_to_corpus(language_mode.clone(), &[ENGLISH_TEXT]).build();
    // Run once beforehand to warm up the cache
    embedder.embed(ENGLISH_TEXT);

    bencher.bench(|| embedder.embed(ENGLISH_TEXT));
}

#[divan::bench(types = [usize, u32, u64], sample_count = 10000)]
fn dimension<T>(bencher: Bencher)
where
    T: TokenEmbedder + Hash + Eq,
{
    bencher
        .with_inputs(|| {
            let embedder =
                EmbedderBuilder::<T>::with_fit_to_corpus(Language::English, &[ENGLISH_TEXT])
                    .build();
            // Run once beforehand to warm up the cache
            embedder.embed(ENGLISH_TEXT);
            embedder
        })
        .bench_values(|embedder| embedder.embed(ENGLISH_TEXT));
}

#[divan::bench]
fn fit_to_corpus(bencher: Bencher) {
    let corpus = [ENGLISH_TEXT; 5000];
    // Run once beforehand to warm up the cache
    let _ = EmbedderBuilder::<u32>::with_fit_to_corpus(Language::English, &corpus).build();

    bencher
        .bench(|| EmbedderBuilder::<u32>::with_fit_to_corpus(Language::English, &corpus).build());
}
