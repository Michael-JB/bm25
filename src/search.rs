use crate::embedder::{
    DefaultEmbeddingDimension, Embedder, EmbedderBuilder, EmbeddingDimension, LanguageMode,
};
use std::{
    collections::{HashMap, HashSet},
    fmt::{self, Debug, Display},
    hash::Hash,
    marker::PhantomData,
};

use crate::embedder::Embedding;

/// A search engine that ranks documents with BM25.
/// K is the type of the document id and D is the type of the embedding dimension.
pub struct SearchEngine<K, D = DefaultEmbeddingDimension>
where
    K: Eq + Hash + Clone,
    D: EmbeddingDimension,
{
    /// The embedder used to convert documents into embeddings.
    embedder: Embedder<D>,
    /// A mapping from document ids to the document embeddings.
    document_embeddings: HashMap<K, Embedding<D>>,
    /// A mapping from document ids to document contents.
    documents: HashMap<K, String>,
    /// A mapping from token indices to the number of documents that contain that token.
    term_frequencies: HashMap<D, u32>,
    /// A mapping from token indices to the set of documents that contain that token.
    inverted_term_index: HashMap<D, HashSet<K>>,
}

impl<K, D> Display for SearchEngine<K, D>
where
    K: Eq + Hash + Clone + Debug,
    D: EmbeddingDimension,
{
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "SearchEngine {{ documents: {:?} }}", self.documents)
    }
}

/// A document that you can insert into a search engine. K is the type of the document, allowing
/// you to use any type as a document id. Note that it is more effient to use a numeric type.
#[derive(Eq, PartialEq, Debug, Clone, PartialOrd, Hash)]
pub struct Document<K>
where
    K: Eq + Hash + Clone + Debug,
{
    /// A unique identifier for the document.
    pub id: K,
    /// The contents of the document.
    pub contents: String,
}

impl<K> Display for Document<K>
where
    K: Eq + Hash + Clone + Debug,
{
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.contents)
    }
}

impl<K> Document<K>
where
    K: Eq + Hash + Clone + Debug,
{
    /// Creates a new document with the given id and contents.
    pub fn new(id: K, contents: impl Into<String>) -> Document<K> {
        Document {
            id,
            contents: contents.into(),
        }
    }
}

/// A search result, containing a document and its BM25 score.
#[derive(PartialEq, Debug)]
pub struct SearchResult<K>
where
    K: Eq + Hash + Clone + Debug,
{
    /// The document that was found.
    pub document: Document<K>,
    /// The BM25 score of the document. A higher score means the document is more relevant to the
    /// query.
    pub score: f32,
}

impl<K, D> SearchEngine<K, D>
where
    K: Eq + Hash + Clone + Debug,
    D: EmbeddingDimension,
{
    /// Upserts a document into the search engine. If a document with the same id already exists,
    /// it will be replaced. Note that upserting a document will change the true value of `avgdl`.
    /// The more `avgdl` drifts from its true value, the less accurate the BM25 scores will be.
    pub fn upsert(&mut self, document: impl Into<Document<K>>) {
        let document = document.into();
        if self.documents.contains_key(&document.id) {
            self.remove(&document.id);
        }
        let Embedding { indices, values } = self.embedder.embed(document.contents.as_str());
        for token_index in indices.iter() {
            let term_frequency = self
                .term_frequencies
                .entry(token_index.clone())
                .or_insert(0);
            *term_frequency += 1;
            let document_set = self
                .inverted_term_index
                .entry(token_index.clone())
                .or_default();
            document_set.insert(document.id.clone());
        }
        self.document_embeddings
            .insert(document.id.clone(), Embedding { indices, values });
        self.documents
            .insert(document.id.clone(), document.contents);
    }

    /// Removes a document from the search engine if it exists.
    pub fn remove(&mut self, document_id: &K) {
        if let Some(document) = self.documents.remove(document_id) {
            let Embedding { indices, .. } = self.embedder.embed(&document);
            for token_index in indices.iter() {
                if let Some(term_frequency) = self.term_frequencies.get_mut(token_index) {
                    *term_frequency -= 1;
                }
                if let Some(document_set) = self.inverted_term_index.get_mut(token_index) {
                    document_set.remove(document_id);
                }
            }
            self.document_embeddings.remove(document_id);
        }
    }

    /// Gets the contents of a document by its id.
    pub fn get(&self, document_id: &K) -> Option<Document<K>> {
        self.documents.get(document_id).map(|contents| Document {
            id: document_id.clone(),
            contents: contents.clone(),
        })
    }

    fn idf(&self, token_index: &D) -> f32 {
        let term_frequency = *self.term_frequencies.get(token_index).unwrap_or(&0) as f32;
        let numerator = self.documents.len() as f32 - term_frequency + 0.5;
        let denominator = term_frequency + 0.5;
        (1f32 + (numerator / denominator)).ln()
    }

    /// Searches the documents for the given query and returns the top `limit` results.
    /// Only the document contents are searched, not the document ids.
    pub fn search(&self, query: &str, limit: impl Into<Option<usize>>) -> Vec<SearchResult<K>> {
        let Embedding { indices, .. } = self.embedder.embed(query);

        // Reduce search space by filtering out all documents whose score would be 0
        let relevent_embedding_it = indices
            .iter()
            .filter_map(|token_index| self.inverted_term_index.get(token_index))
            .flat_map(|document_set| document_set.iter())
            .collect::<HashSet<_>>()
            .into_iter()
            .filter_map(|document_id| {
                self.document_embeddings
                    .get(document_id)
                    .map(|embedding| (document_id, embedding))
            });

        // Compute the score of each candidate document
        let mut scores = HashMap::new();
        for (document_id, embedding) in relevent_embedding_it {
            let mut document_score = 0f32;
            for token_index in indices.iter() {
                let token_idf = self.idf(token_index);
                let token_index_value = embedding
                    .indices
                    .iter()
                    .zip(embedding.values.iter())
                    .find(|(i, _)| **i == *token_index)
                    .map(|(_, v)| *v)
                    .unwrap_or(0f32);
                let token_score = token_idf * token_index_value;
                document_score += token_score;
            }
            scores.insert(document_id, document_score);
        }

        // Sort and format results
        let mut results: Vec<_> = scores.iter().collect();
        results.sort_by(|a, b| b.1.partial_cmp(a.1).unwrap());
        results
            .into_iter()
            .take(limit.into().unwrap_or(usize::MAX))
            .filter_map(|(id, score)| {
                self.get(id).map(|document| SearchResult {
                    document,
                    score: *score,
                })
            })
            .collect()
    }
}

/// A consuming builder for SearchEngine.
pub struct SearchEngineBuilder<K, D = DefaultEmbeddingDimension>
where
    K: Eq + Hash + Clone + Debug,
    D: EmbeddingDimension,
{
    embedder_builder: EmbedderBuilder<D>,
    documents: Vec<Document<K>>,
    document_id_type: PhantomData<K>,
    embedding_dimension_type: PhantomData<D>,
}

impl<K, D> SearchEngineBuilder<K, D>
where
    K: Eq + Hash + Clone + Debug,
    D: EmbeddingDimension,
{
    /// Constructs a new SearchEngineBuilder with the given average document length. Use this if you
    /// know the average document length in advance. If you don't, but you have your full corpus
    /// ahead of time, use `with_documents` or `with_corpus` instead.
    ///
    /// If you have neither the full corpus nor a sample of it, you can configure the embedder to
    /// disregard document length by setting `b` to 0.0. In this case, it doesn't matter what
    /// value you pass to `with_avgdl`.
    ///
    /// The average document length is the average number of tokens in a document from your corpus;
    /// if you need access to this value, you can construct an Embedder and call `avgdl` on it.
    pub fn with_avgdl(avgdl: f32) -> SearchEngineBuilder<K, D> {
        SearchEngineBuilder {
            embedder_builder: EmbedderBuilder::<D>::with_avgdl(avgdl),
            documents: Vec::new(),
            document_id_type: PhantomData,
            embedding_dimension_type: PhantomData,
        }
    }

    /// Constructs a new SearchEngineBuilder with the given documents. The search engine will fit
    /// to the given documents, using the given language mode. When you call `build`, the builder
    /// will pre-populate the search engine with the given documents, and set its language mode to
    /// `language_mode`.
    pub fn with_documents(
        language_mode: impl Into<LanguageMode>,
        documents: impl IntoIterator<Item = impl Into<Document<K>>>,
    ) -> SearchEngineBuilder<K, D> {
        let documents = documents.into_iter().map(|d| d.into()).collect::<Vec<_>>();
        SearchEngineBuilder {
            embedder_builder: EmbedderBuilder::<D>::with_fit_to_corpus(
                language_mode,
                &documents
                    .iter()
                    .map(|d| d.contents.as_str())
                    .collect::<Vec<_>>(),
            ),
            documents,
            document_id_type: PhantomData,
            embedding_dimension_type: PhantomData,
        }
    }

    /// Constructs a new SearchEngineBuilder with the corpus. The search engine will fit
    /// to the given corpus, using the given language mode. When you call `build`, the builder
    /// will pre-populate the search engine with the given corpus. This function will
    /// automatically generate u32 ids for each entry in your corpus.
    pub fn with_corpus(
        language_mode: impl Into<LanguageMode>,
        corpus: impl IntoIterator<Item = impl Into<String>>,
    ) -> SearchEngineBuilder<u32, D> {
        let documents = corpus
            .into_iter()
            .enumerate()
            .map(|(id, document)| Document::new(id as u32, document.into()))
            .collect::<Vec<_>>();
        SearchEngineBuilder::<u32, D>::with_documents(language_mode.into(), documents)
    }

    /// Sets the language mode of the embedder.
    pub fn language_mode(mut self, language_mode: LanguageMode) -> SearchEngineBuilder<K, D> {
        self.embedder_builder.language_mode(language_mode);
        self
    }

    /// Sets the k1 parameter of the embedder.
    pub fn k1(mut self, k1: f32) -> SearchEngineBuilder<K, D> {
        self.embedder_builder.k1(k1);
        self
    }

    /// Sets the b parameter of the embedder.
    pub fn b(mut self, b: f32) -> SearchEngineBuilder<K, D> {
        self.embedder_builder.b(b);
        self
    }

    /// Overrides the average document length of the embedder.
    pub fn avgdl(mut self, avgdl: f32) -> SearchEngineBuilder<K, D> {
        self.embedder_builder.avgdl(avgdl);
        self
    }

    /// Builds the search engine.
    pub fn build(self) -> SearchEngine<K, D> {
        let mut search_engine = SearchEngine::<K, D> {
            embedder: self.embedder_builder.build(),
            document_embeddings: HashMap::new(),
            inverted_term_index: HashMap::new(),
            term_frequencies: HashMap::new(),
            documents: HashMap::new(),
        };
        for document in self.documents {
            search_engine.upsert(document);
        }
        search_engine
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{
        test_data_loader::tests::{read_recipes, Recipe},
        Language,
    };

    impl From<Recipe> for Document<String> {
        fn from(value: Recipe) -> Self {
            Document::new(value.title, value.recipe)
        }
    }

    fn create_recipe_search_engine(
        language_mode: impl Into<LanguageMode>,
    ) -> SearchEngine<String, u32> {
        let recipes = read_recipes("recipes_en.csv");

        SearchEngineBuilder::with_documents(language_mode, recipes).build()
    }

    #[test]
    fn search_returns_relevant_documents() {
        let corpus = vec!["space station", "bacon and avocado sandwich"];
        let search_engine =
            SearchEngineBuilder::<u32>::with_corpus(Language::English, corpus).build();

        let results = search_engine.search("sandwich with bacon", 5);
        assert!(results.len() == 1);
        assert!(results[0].document.contents == "bacon and avocado sandwich");
        assert!(results[0].score > 0.0);
    }

    #[test]
    fn search_does_not_return_unrelated_documents() {
        let corpus = vec!["space station", "bacon and avocado sandwich"];
        let search_engine =
            SearchEngineBuilder::<u32>::with_corpus(Language::English, corpus).build();

        let results = search_engine.search("maths and computer science", 5);
        assert!(results.is_empty());
    }

    #[test]
    fn it_can_insert_a_document() {
        let mut search_engine = SearchEngineBuilder::<&str>::with_avgdl(2.0).build();
        let document = Document::new("hello world", "bananas and apples");
        let document_id = document.id;

        search_engine.upsert(document.clone());
        let result = search_engine.get(&document_id);

        assert!(result.unwrap() == document);
    }

    #[test]
    fn it_can_remove_a_document() {
        let mut search_engine = SearchEngineBuilder::<usize>::with_avgdl(2.0).build();
        let document = Document::new(123, "bananas and apples");
        let document_id = document.id.clone();

        search_engine.upsert(document);
        search_engine.remove(&document_id);

        assert!(search_engine.get(&document_id).is_none());
    }

    #[test]
    fn it_can_update_a_document() {
        let document_id = "hello_world";
        let document = Document::new(document_id, "bananas and apples");
        let mut search_engine =
            SearchEngineBuilder::<&str>::with_documents(Language::English, vec![document]).build();
        let new_document = Document::new(document_id, "oranges and papayas");

        search_engine.upsert(new_document.clone());
        let result = search_engine.get(&document_id);

        assert!(result.unwrap() == new_document);
    }

    #[test]
    fn handles_empty_input() {
        let mut search_engine = SearchEngineBuilder::<u32>::with_avgdl(2.0).build();
        let document = Document::new(123, "");

        search_engine.upsert(document);

        let results = search_engine.search("bacon sandwich", 5);
        assert!(results.is_empty());
    }

    #[test]
    fn handles_empty_search() {
        let mut search_engine = SearchEngineBuilder::<u32>::with_avgdl(2.0).build();
        let document = Document::new(123, "pencil and paper");

        search_engine.upsert(document);

        let results = search_engine.search("", 5);
        assert!(results.is_empty());
    }

    #[test]
    fn it_returns_exact_matches_with_highest_score() {
        let search_engine = create_recipe_search_engine(LanguageMode::Fixed(Language::English));

        let results = search_engine.search(
            "To make guacamole, start by mashing 2 ripe avocados in a bowl.",
            None,
        );

        assert!(!results.is_empty());
        assert_eq!(results[0].document.id, "Guacamole");
    }

    #[test]
    fn it_only_returns_results_containing_query() {
        let search_engine = create_recipe_search_engine(Language::English);

        let results = search_engine.search("vegetable", 5);

        // At least 5 recipes contain the word "vegetable"
        assert_eq!(results.len(), 5);
        assert!(results
            .iter()
            .all(|result| result.document.contents.contains("vegetable")));
    }

    #[test]
    fn it_returns_results_sorted_by_score() {
        let search_engine = create_recipe_search_engine(Language::English);

        let results = search_engine.search("chicken", 1000);

        assert!(!results.is_empty());
        assert!(results
            .windows(2)
            .all(|result_pair| { result_pair[0].score >= result_pair[1].score }));
    }

    #[test]
    fn it_ranks_shorter_documents_higher() {
        let documents = [
            Document {
                id: 0,
                contents: "Correct horse battery staple bacon bacon bacon".to_string(),
            },
            Document {
                id: 1,
                contents: "Correct horse battery staple".to_string(),
            },
        ];
        let search_engine =
            SearchEngineBuilder::<u32>::with_documents(Language::English, documents).build();

        let results = search_engine.search("staple", 2);

        assert_eq!(results.len(), 2);
        assert_eq!(results[0].document.id, 1);
        assert_eq!(results[1].document.id, 0);
        assert!(results[0].score > results[1].score);
    }
}
