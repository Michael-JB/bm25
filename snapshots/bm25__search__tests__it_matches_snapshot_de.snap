---
source: src/search.rs
expression: results
---
[
    SearchResult {
        document: Document {
            id: "Apfelkuchen",
            contents: "Um Apfelkuchen zu machen, heizen Sie Ihren Ofen auf 220°C (425°F) vor. In einer großen Schüssel mischen Sie 6 geschälte und in Scheiben geschnittene Äpfel mit 1/2 Tasse Zucker, 1 Teelöffel Zimt und 1/4 Teelöffel Muskatnuss. Die Äpfel gleichmäßig bedecken. Rollen Sie einen Teigboden aus und legen Sie ihn in eine 9-Zoll-Pie-Form. Gießen Sie die Apfelmischung in den Teig und geben Sie 2 Esslöffel Butter darauf. Rollen Sie den zweiten Teigboden aus und legen Sie ihn über die Äpfel, drücken Sie die Ränder zusammen, um ihn zu verschließen. Schneiden Sie einige Schlitze in den oberen Teig, um Dampf entweichen zu lassen. Backen Sie den Kuchen 45-50 Minuten lang, bis die Kruste goldbraun und die Füllung blubbert. Lassen Sie den Kuchen abkühlen, bevor Sie ihn servieren.",
        },
        score: 0.9877103,
    },
    SearchResult {
        document: Document {
            id: "Auberginen-Parmesan",
            contents: "Um Auberginen-Parmesan zuzubereiten, heizen Sie Ihren Ofen auf 190°C (375°F) vor. Schneiden Sie 2 mittelgroße Auberginen in 1/4 Zoll dicke Scheiben. Würzen Sie die Scheiben mit Salz und lassen Sie sie 30 Minuten stehen, um überschüssige Feuchtigkeit zu entziehen. Tupfen Sie die Auberginenscheiben mit Papiertüchern trocken. Wenden Sie jede Scheibe in Mehl, tauchen Sie sie in geschlagene Eier und bestreichen Sie sie mit Semmelbröseln. In einer großen Pfanne erhitzen Sie 2 Esslöffel Olivenöl bei mittlerer Hitze und braten Sie die Auberginenscheiben goldbraun an beiden Seiten. In einer Auflaufform schichten Sie die gebratenen Auberginenscheiben mit Marinara-Sauce und geriebenem Mozzarella-Käse. Wiederholen Sie die Schichten, bis alle Zutaten aufgebraucht sind, und enden Sie mit einer Schicht Käse. Backen Sie das Gericht 25-30 Minuten, bis der Käse geschmolzen und blubbert. Heiß servieren.",
        },
        score: 0.9648115,
    },
    SearchResult {
        document: Document {
            id: "BBQ-Hähnchenflügel",
            contents: "Um BBQ-Hähnchenflügel zuzubereiten, heizen Sie Ihren Ofen auf 200°C (400°F) vor. Geben Sie 900 g Hähnchenflügel in eine große Schüssel und vermengen Sie sie mit 2 Esslöffeln Pflanzenöl, Salz und Pfeffer. Verteilen Sie die Flügel auf einem mit Backpapier ausgelegten Backblech. Backen Sie die Flügel im vorgeheizten Ofen 25-30 Minuten lang, wobei Sie sie zur Hälfte wenden, bis die Flügel goldbraun und knusprig sind. Nehmen Sie die Flügel aus dem Ofen und wenden Sie sie in Ihrer Lieblings-BBQ-Sauce. Legen Sie die Flügel für weitere 5 Minuten in den Ofen, um die Sauce festzusetzen. Heiß servieren, mit zusätzlicher Sauce an der Seite.",
        },
        score: 1.0842069,
    },
    SearchResult {
        document: Document {
            id: "Bananenbrot",
            contents: "Um Bananenbrot zu machen, heizen Sie den Ofen auf 175°C vor. In einer großen Schüssel 3 reife Bananen mit einer Gabel zerdrücken. Rühren Sie 1/3 Tasse geschmolzene Butter, 1/2 Tasse Zucker, 1 geschlagenes Ei und 1 Teelöffel Vanilleextrakt ein. In einer separaten Schüssel 1 1/2 Tassen Mehl, 1 Teelöffel Natron und eine Prise Salz vermengen. Die trockenen Zutaten nach und nach zu den nassen Zutaten geben und rühren, bis sie gerade so vermischt sind. Den Teig in eine gefettete Kastenform gießen und 60 Minuten backen, oder bis ein Zahnstocher, der in die Mitte gesteckt wird, sauber herauskommt. Lassen Sie das Brot 10 Minuten in der Form abkühlen, bevor Sie es auf ein Kuchengitter legen, um es vollständig abzukühlen.",
        },
        score: 0.99957216,
    },
    SearchResult {
        document: Document {
            id: "Blaubeermuffins",
            contents: "Um Blaubeermuffins zu machen, heizen Sie Ihren Ofen auf 190°C vor und legen Sie eine Muffinform mit Papierförmchen aus. In einer großen Schüssel 1 1/2 Tassen Mehl, 3/4 Tasse Zucker, 1/2 Teelöffel Salz und 2 Teelöffel Backpulver vermengen. In einer separaten Schüssel 1/3 Tasse Pflanzenöl, 1 großes Ei und 1/3 Tasse Milch vermischen. Die nassen Zutaten zu den trockenen Zutaten geben und rühren, bis sie gerade so vermischt sind. Vorsichtig 1 Tasse frische oder gefrorene Blaubeeren unterheben. Den Teig gleichmäßig in die Muffinform verteilen und 20-25 Minuten backen, oder bis ein Zahnstocher in der Mitte sauber herauskommt. Lassen Sie die Muffins einige Minuten in der Form abkühlen, bevor Sie sie auf ein Gitter legen, um sie vollständig abzukühlen.",
        },
        score: 0.9648115,
    },
    SearchResult {
        document: Document {
            id: "Gefüllte Paprika",
            contents: "Um gefüllte Paprika zu machen, heizen Sie Ihren Ofen auf 190°C vor. Schneiden Sie die Oberseiten von 4 Paprikaschoten ab und entfernen Sie die Kerne. In einer großen Pfanne 1 Esslöffel Olivenöl bei mittlerer Hitze erhitzen. Fügen Sie 1 gewürfelte Zwiebel und 2 gehackte Knoblauchzehen hinzu und braten Sie, bis sie weich sind. Geben Sie 450 g Hackfleisch hinzu und braten Sie es, bis es braun ist. Überschüssiges Fett abgießen. 1 Tasse gekochten Reis, 1 Dose gewürfelte Tomaten, Salz, Pfeffer und 1 Teelöffel italienische Gewürze einrühren. Jede Paprika mit der Fleischmischung füllen und in eine Auflaufform stellen. Mit geriebenem Käse bestreuen und mit Folie abdecken. 25-30 Minuten backen, dann die Folie entfernen und weitere 10 Minuten backen, bis der Käse goldbraun ist.",
        },
        score: 1.3532338,
    },
    SearchResult {
        document: Document {
            id: "Hähnchen-Parmesan",
            contents: "Um Hähnchen-Parmesan zu machen, heizen Sie Ihren Ofen auf 200°C vor. Würzen Sie 2 Hähnchenbrustfilets ohne Haut und Knochen mit Salz und Pfeffer. Wälzen Sie jedes Hähnchenbrustfilet in Mehl, tauchen Sie es dann in geschlagene Eier und schließlich in Semmelbrösel, die mit geriebenem Parmesan gemischt wurden. In einer großen Pfanne 2 Esslöffel Olivenöl bei mittlerer Hitze erhitzen. Braten Sie das Hähnchen 3-4 Minuten auf jeder Seite, bis es goldbraun ist. Legen Sie das Hähnchen in eine Auflaufform und bedecken Sie es mit Marinara-Sauce und geriebenem Mozzarella. 15-20 Minuten backen, bis der Käse geschmolzen und sprudelnd ist. Heiß servieren, über Spaghetti oder Ihrer Lieblingspasta.",
        },
        score: 1.0842069,
    },
    SearchResult {
        document: Document {
            id: "Knoblauchbrot",
            contents: "Um Knoblauchbrot zu machen, heizen Sie Ihren Ofen auf 175°C (350°F) vor. In einer kleinen Schüssel mischen Sie 1/2 Tasse weiche Butter, 3 gehackte Knoblauchzehen und 2 Esslöffel gehackte frische Petersilie. Schneiden Sie ein Baguette der Länge nach in zwei Hälften. Bestreichen Sie die Schnittseiten des Brotes gleichmäßig mit der Knoblauchbutter-Mischung. Legen Sie die Brothälften auf ein Backblech und backen Sie sie 10-12 Minuten lang, bis die Ränder goldbraun und knusprig sind. Für extra käsiges Knoblauchbrot streuen Sie geriebenen Mozzarella oder Parmesan über das Brot und backen es weitere 3-5 Minuten, bis der Käse geschmolzen und blubbernd ist. Schneiden und warm servieren.",
        },
        score: 1.4410969,
    },
    SearchResult {
        document: Document {
            id: "Kürbiskuchen",
            contents: "Um Kürbiskuchen zu machen, heizen Sie Ihren Ofen auf 220°C (425°F) vor. In einer großen Schüssel verquirlen Sie 2 Tassen Kürbispüree aus der Dose, 1 Dose gesüßte Kondensmilch, 2 große Eier, 1 Teelöffel Zimt, 1/2 Teelöffel gemahlenen Ingwer, 1/4 Teelöffel gemahlene Nelken und 1/2 Teelöffel Salz. Gießen Sie die Mischung in eine ungebackene 23 cm (9-Zoll) Kuchenkruste. Backen Sie den Kuchen 15 Minuten, reduzieren Sie dann die Ofentemperatur auf 175°C (350°F) und backen Sie weitere 35-40 Minuten, oder bis ein Messer in der Mitte sauber herauskommt. Lassen Sie den Kuchen vollständig abkühlen, bevor Sie ihn mit Schlagsahne servieren.",
        },
        score: 1.4046171,
    },
    SearchResult {
        document: Document {
            id: "Lasagne",
            contents: "Um Lasagne zuzubereiten, heizen Sie den Ofen auf 190°C (375°F) vor. Kochen Sie 12 Lasagne-Nudeln nach Packungsanweisung. In einer großen Pfanne braten Sie 1 Pfund Hackfleisch mit 1 gewürfelten Zwiebel und 2 gehackten Knoblauchzehen an. Lassen Sie das überschüssige Fett ab und rühren Sie 1 Glas Marinara-Sauce ein, lassen Sie es 10 Minuten köcheln. In einer separaten Schüssel mischen Sie 1 Tasse Ricotta-Käse mit 1 geschlagenen Ei und 1/2 Tasse geriebenem Parmesan. Zum Zusammenstellen der Lasagne verteilen Sie eine dünne Schicht Fleischsauce auf dem Boden einer Auflaufform. Schichten Sie 3 Nudeln darauf, dann ein Drittel der Ricotta-Mischung und geriebenen Mozzarella-Käse. Wiederholen Sie die Schichten und enden Sie mit einer letzten Schicht Fleischsauce und Mozzarella. Decken Sie die Form mit Folie ab und backen Sie sie 25 Minuten, dann nehmen Sie die Folie ab und backen Sie weitere 15 Minuten, bis der Käse geschmolzen und blubbernd ist. Lassen Sie die Lasagne 10 Minuten abkühlen, bevor Sie sie schneiden und servieren.",
        },
        score: 1.2329572,
    },
    SearchResult {
        document: Document {
            id: "Schokoladenkekse",
            contents: "Um klassische Schokoladenkekse zu machen, heizen Sie den Ofen auf 175°C (350°F) vor. In einer großen Schüssel verrühren Sie 1 Tasse weiche Butter, 1 Tasse weißen Zucker und 1 Tasse braunen Zucker, bis die Masse glatt ist. Schlagen Sie 2 Eier einzeln unter, dann rühren Sie 2 Teelöffel Vanilleextrakt ein. In einer separaten Schüssel vermischen Sie 3 Tassen Mehl, 1 Teelöffel Natron und 1/2 Teelöffel Salz. Nach und nach mischen Sie die trockenen Zutaten unter die feuchte Mischung. Heben Sie 2 Tassen halbsüße Schokoladenchips unter. Geben Sie Esslöffel große Teigbällchen auf ungefettete Backbleche. Backen Sie sie 10-12 Minuten lang, bis die Ränder goldbraun sind. Lassen Sie die Kekse einige Minuten auf dem Backblech abkühlen, bevor Sie sie auf ein Kuchengitter legen, um sie vollständig abzukühlen.",
        },
        score: 0.9648115,
    },
    SearchResult {
        document: Document {
            id: "Schokoladenkuchen",
            contents: "Um einen saftigen Schokoladenkuchen zu machen, heizen Sie Ihren Ofen auf 175°C (350°F) vor. Fetten und mehlen Sie zwei 23 cm (9-Zoll) runde Kuchenformen. In einer großen Schüssel verrühren Sie 2 Tassen Zucker, 1 3/4 Tassen Mehl, 3/4 Tasse Kakaopulver, 1 1/2 Teelöffel Backpulver, 1 1/2 Teelöffel Natron und 1 Teelöffel Salz. Fügen Sie 2 Eier, 1 Tasse Milch, 1/2 Tasse Pflanzenöl und 2 Teelöffel Vanilleextrakt hinzu. Schlagen Sie die Mischung 2 Minuten lang bei mittlerer Geschwindigkeit. Rühren Sie 1 Tasse kochendes Wasser ein (der Teig wird dünnflüssig). Gießen Sie den Teig gleichmäßig in die vorbereiteten Formen. Backen Sie 30-35 Minuten oder bis ein Zahnstocher in der Mitte sauber herauskommt. Lassen Sie die Kuchen 10 Minuten in den Formen abkühlen, dann auf einem Kuchengitter vollständig abkühlen, bevor Sie sie glasieren.",
        },
        score: 0.87366533,
    },
    SearchResult {
        document: Document {
            id: "Überbackene Ziti",
            contents: "Um überbackene Ziti zuzubereiten, heizen Sie den Ofen auf 190°C (375°F) vor. Kochen Sie 450 g Ziti-Nudeln nach Packungsanweisung, lassen Sie sie abtropfen und stellen Sie sie beiseite. In einer großen Pfanne braten Sie 450 g italienische Wurst bei mittlerer Hitze an und zerbröckeln sie in kleine Stücke. Rühren Sie 1 Glas Marinara-Sauce ein und lassen Sie es 5 Minuten köcheln. In einer großen Schüssel mischen Sie die gekochten Ziti mit der Wurstmischung, 1 Tasse Ricotta-Käse und 1/2 Tasse geriebenem Parmesan. Geben Sie die Mischung in eine gefettete Auflaufform und bestreuen Sie sie mit 2 Tassen geriebenem Mozzarella. Backen Sie 25-30 Minuten, bis der Käse geschmolzen und blubbernd ist. Lassen Sie die Ziti einige Minuten abkühlen, bevor Sie sie servieren.",
        },
        score: 0.97612655,
    },
]
